### PR TITLE
Move volumeBalance to MultipahseporomechanicsKernel.

### PR DESCRIPTION
**Pull request by CusiniM**
rebaseline: https://github.com/GEOSX/integratedTests/pull/140

### CODE REVIEW

**Comment by CusiniM** [comment]
Can we get this in quickly?

**Review by klevzoff — APPROVED** [review]
All good, just minor suggestions

As an aside, I've noticed that this is repeating the same calculation at every quadrature point, with the difference being `porosityNew` which varies from point to point. Makes me wonder if all of the volume balance stuff (except the porosity multiplier) should be computed outside of the quadrature loop, then added with an integrated porosity multiplier. But this is a topic for another day perhaps.

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 1]
```diff
@@ -215,6 +217,8 @@ class Multiphase :
     real64 localDispFlowJacobian[numDispDofPerElem][numMaxComponents + 1];
     real64 localFlowDispJacobian[numMaxComponents][numDispDofPerElem];
     real64 localFlowFlowJacobian[numMaxComponents][numMaxComponents + 1];
+    real64 localVolBalanceResidual;
+    real64 localVolBalanceJacobian[numMaxComponents + 1];
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:221]:
```suggestion
    real64 localVolBalanceResidual[1];
    real64 localVolBalanceJacobian[1][numMaxComponents + 1];
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 2]
```diff
@@ -194,6 +194,8 @@ class Multiphase :
       localDispFlowJacobian{ {0.0} },
       localFlowDispJacobian{ {0.0} },
       localFlowFlowJacobian{ {0.0} },
+      localVolBalanceResidual( 0.0 ),
+      localVolBalanceJacobian{ 0 },
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:198]:
```suggestion
      localVolBalanceResidual{ 0 },
      localVolBalanceJacobian{ {0.0} },
```

For consistency with what is done in the single-phase poromechanics kernel (see also suggestions that that follow).

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 3]
```diff
@@ -429,6 +433,21 @@ class Multiphase :
         stack.localFlowDispJacobian[ic][a*3+2] += dPorosity_dVolStrainIncrement * componentAmount[ic] * dNdX[a][2] * detJxW;
       }
     }
+
+    // --- Volume balance equation
+    // sum contributions to component accumulation from each phase
+    stack.localVolBalanceResidual += porosityNew * detJxW;
+    for( localIndex ip = 0; ip < NP; ++ip )
+    {
+      stack.localVolBalanceResidual -= m_fluidPhaseSaturation( k, ip ) * porosityNew * detJxW;
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:442]:
```suggestion
      stack.localVolBalanceResidual[0] -= m_fluidPhaseSaturation( k, ip ) * porosityNew * detJxW;
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 4]
```diff
@@ -429,6 +433,21 @@ class Multiphase :
...
+      stack.localVolBalanceJacobian[0] -=
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:443]:
```suggestion
      stack.localVolBalanceJacobian[0][0] -=
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 5]
```diff
@@ -429,6 +433,21 @@ class Multiphase :
...
+        ( m_dFluidPhaseSaturation_dPressure( k, ip ) * porosityNew + dPorosity_dPressure * m_fluidPhaseSaturation( k, ip ) ) * detJxW;
+
+      for( localIndex jc = 0; jc < NC; ++jc )
+      {
+        stack.localVolBalanceJacobian[jc+1] -= m_dFluidPhaseSaturation_dGlobalCompDensity( k, ip, jc )  * porosityNew * detJxW;
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:448]:
```suggestion
        stack.localVolBalanceJacobian[0][jc+1] -= m_dFluidPhaseSaturation_dGlobalCompDensity( k, ip, jc )  * porosityNew * detJxW;
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 6]
```diff
@@ -488,6 +508,18 @@ class Multiphase :
       }
     }
 
+    localIndex const volBalanceRow = LvArray::integerConversion< localIndex >( stack.localFlowDofIndex[0] + m_numComponents - m_dofRankOffset );
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:511]:
```suggestion
    dof = LvArray::integerConversion< localIndex >( stack.localFlowDofIndex[0] + m_numComponents - m_dofRankOffset );
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 7]
```diff
@@ -488,6 +508,18 @@ class Multiphase :
...
+    if( 0 <= volBalanceRow && volBalanceRow < m_matrix.numRows() )
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:512]:
```suggestion
    if( 0 <= dof && dof < m_matrix.numRows() )
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 8]
```diff
@@ -488,6 +508,18 @@ class Multiphase :
...
+    {
+
+      m_matrix.template addToRowBinarySearchUnsorted< serialAtomic >( volBalanceRow,
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:515]:
```suggestion
      m_matrix.template addToRowBinarySearchUnsorted< serialAtomic >( dof,
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsSolver.cpp [hunk 9]
```diff
@@ -210,14 +210,6 @@ void MultiphasePoromechanicsSolver::assembleSystem( real64 const time_n,
...
                                    localMatrix,
                                    localRhs );
-
-  // Cell-based contribution
-  // TODO: move in PoroelasticKernels
-  m_flowSolver->assembleVolumeBalanceTerms( domain,
-                                            dofManager,
-                                            localMatrix,
-                                            localRhs );
-
```
Comment (rrsettgast) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsSolver.cpp:220]: I don't see the deletion of the actual function anywhere?

**Review by rrsettgast — APPROVED** [review]

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 10]
```diff
@@ -488,6 +508,18 @@ class Multiphase :
...
+                                                                      stack.localFlowDofIndex,
+                                                                      stack.localVolBalanceJacobian,
+                                                                      m_numComponents + 1 );
+
+      m_rhs[volBalanceRow] += stack.localVolBalanceResidual;
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:520]:
```suggestion
      RAJA::atomicAdd< serialAtomic >( &m_rhs[dof], stack.localVolBalanceResidual[0] );
```

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsSolver.cpp (hunk 9 above)
Comment (CusiniM) [reply src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsSolver.cpp:220]: it's still used for flow only simulations.

File: src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp [hunk 11]
```diff
@@ -488,6 +508,18 @@ class Multiphase :
       }
     }
 
+    localIndex const volBalanceRow = LvArray::integerConversion< localIndex >( stack.localFlowDofIndex[0] + m_numComponents - m_dofRankOffset );
+    if( 0 <= volBalanceRow && volBalanceRow < m_matrix.numRows() )
+    {
+
+      m_matrix.template addToRowBinarySearchUnsorted< serialAtomic >( volBalanceRow,
+                                                                      stack.localFlowDofIndex,
+                                                                      stack.localVolBalanceJacobian,
```
Comment (castelletto1) [inline src/coreComponents/physicsSolvers/multiphysics/MultiphasePoromechanicsKernel.hpp:517]:
```suggestion
                                                                      stack.localVolBalanceJacobian[0],
```

**Review by castelletto1 — APPROVED** [review]
Thank you for taking of this! I just made a few suggestions for consistency with what we have in the single-phase poromechanics kernel.

**Pull request merged**